# [자연어처리]
# 6주차(6-2). 차원 축소





# 201801544 김진원 <- 본인의 학번, 이름으로 변경하여 제출
* **모든 셀 실행** 후 제출하시기 바랍니다.
* **실습 (6-2-1, 6-2-2)**이 있습니다. (제출 기한: 10/11(수) 23시 59분까지)

# 이제 Colab 사용에 모든 수강생분들이 어느정도 익숙해 졌을 거라 생각됩니다!
> ##### 중간 중간 **'### ... 이 부분을 완성하시오'** 라는 부분의 코드를 완성해야 합니다.





## 5주차 수업 내용 리뷰 & 정규 표현식 마무리 -시작-

In [1]:
import re

m = re.match('[a-z]+', "python") # 소문자가 한번 이상 등장한 패턴을 전부 찾음
print(m.group())

python


In [2]:
# re.compile() 이용하여 작성해보기
# import re는 다시 선언해주지 않아도 되나, 이해를 돕기 위해 매 셀에서 작성되어 있음
import re

pattern = re.compile('[a-z]+')
m = pattern.match("python")
print(m.group())

python


In [3]:
s = pattern.sub("***", "python")
print(s)

***


In [4]:
 s = pattern.sub("***", "python programming")
 print(s)

*** ***


In [5]:
text = '''오늘의 날씨는 전국적으로 맑을 예정입니다. 시원한 가을의 날씨를 느껴보세요.

홍길동 빠른뉴스 기자(gildong@fastnews.com)'''

In [6]:
# 괄호로 둘러싸인 이메일 주소 패턴을 정규 표현식으로 찾아내어 삭제하기
import re

pattern = '\([A-Za-z0-9_\.]+@[A-Za-z\.]+[A-Za-z]\)'
re.sub(pattern, "", text)

'오늘의 날씨는 전국적으로 맑을 예정입니다. 시원한 가을의 날씨를 느껴보세요.\n\n홍길동 빠른뉴스 기자'

In [7]:
# 이메일 주소의 마지막은 com, org, edu, net, co.kr 중 하나라고 가정한다면
import re

pattern = '\([A-Za-z0-9_\.]+@[A-Za-z]+\.(com|org|edu|net|co.kr)\)'
re.sub(pattern, "", text)

'오늘의 날씨는 전국적으로 맑을 예정입니다. 시원한 가을의 날씨를 느껴보세요.\n\n홍길동 빠른뉴스 기자'

In [8]:
# re.compile() 이용하여 작성해보기
import re

# 위의 코드를 re.compile() 이용하여 작성해보시오
### ... 이 부분을 완성하시오
### ... 이 부분을 완성하시오

pattern = re.compile('\([A-Za-z0-9_\.]+@[A-Za-z]+\.(com|org|edu|net|co.kr)\)')
pattern.sub("", text)

'오늘의 날씨는 전국적으로 맑을 예정입니다. 시원한 가을의 날씨를 느껴보세요.\n\n홍길동 빠른뉴스 기자'

In [9]:
import re

text = "안녕하세요, 1234는 숫자이고 5678도 숫자입니다."
pattern = r'\d+'  # 하나 이상의 연속된 숫자에 대한 패턴
numbers = re.findall(pattern, text)
print(numbers)  # ['1234', '5678']

['1234', '5678']


In [10]:
import re

# 정규 표현식을 컴파일하여 pattern 객체 생성
text = "안녕하세요, 1234는 숫자이고 5678도 숫자입니다."
pattern = re.compile(r'\d+')
numbers = pattern.findall(text)
print(numbers)  # ['1234', '5678']

['1234', '5678']


In [11]:
# 정규 표현식을 이용한 토크나이저
# A RegexpTokenizer splits a string into substrings using a regular expression.
from nltk.tokenize import RegexpTokenizer

text1 = "Sorry, I can't go there."
# 문자, 숫자, 언더바(_), 아포스트로피(＇)로 이루어진 3자 이상의 단어로 토크나이즈
tokenizer = RegexpTokenizer("[\w']{3,}") ### ... 이 부분을 완성하시오 # ???는 tokenizer 이름
print(tokenizer.tokenize(text1.lower())) # 소문자로 바꾸고 정규식 적용

['sorry', "can't", 'there']


# [[실습 6-2-1]] 이메일 주소를 찾는 정규 표현식
* 정규 표현식의 패턴은 다양하게 작성될 수 있습니다. 이번에는 이메일 주소를 분류하는 정규표현식을 **\w**를 사용하여 작성해 보시오. (\w를 포함하기만 하면 되고, 그 외에는 자유)
1. 여러분이 생각하는 이메일 주소의 **규칙**을 적고, 그에 대한 정규 표현식을 만들어 보세요.
> 예시) 이메일 주소는 id@domain와 같이 이루어져 있음 <br>
> * 규칙1. id는 길이가 10이하여야 하고, 문자(대문자, 소문자), 숫자, _로만 이루어져야 함 <br>
> * 규칙2. domin은 naver.com, inu.ac.kr, cse.inu.ac.kr과 같이 .을 단위로 2개~4개까지 이루어질 수 있음 <br>
#### **[유의사항] 여러분만의 이메일 주소 '규칙1~규칙N'을 작성하여 제출 바랍니다. (위에 작성된 것은 참고용 예시임)**
2. 규칙을 만족하는/만족하지 않는 테스트 **예시 문장**을 만들고, **예상 결과**와 같이 잘 동작하는지 확인해 보세요.
> 예시) <br>
> 예시 문장) text = '홍길동(gildong2da@inu.ac.kr), 나학생(i_am_a_student@naver.com)' <br>
> 예상 결과) ['gildong2da@inu.ac.kr', 'i_am_a_student@naver.com']
#### **[유의사항] 여러분만의 '예시 문장'과 '예상 결과'를 작성하여 제출 바랍니다. (위에 작성된 것은 참고용 예시임)**
3. 위의 1, 2를 이용하여 코드를 작성하세요.
> 위의 숫자를 찾는 코드를 참고하여 작성하면 도움이 될 겁니다.
```
import re
text = "안녕하세요, 1234는 숫자이고 5678도 숫자입니다."
pattern = re.compile(r'\d+')
numbers = pattern.findall(text)
print(numbers)  # ['1234', '5678']
```






규칙1. ID는 길이가 12이하여야 하고 영어 대소문자와 숫자,
 _만 포함

규칙2. "@" 기호가 입력되어야함.

규칙3. 도메인 이름은 영문자나 숫자로 이루어져야함.

규칙4. 도메인 이름은 최소 하나의 .을 포함해야함.

In [12]:
import re

text = "김진원(a01020532509@gmail.com), 나학생(i_student@inu.ac.kr), 김진영(kjy@cse.inu.ac.kr), 박수민(psm@navercom)"
pattern = re.compile(r'\b\w{1,12}@[a-zA-Z0-9]+\.[a-zA-Z]{2,}(?:\.[a-zA-Z]{2,})*\b')
email = re.findall(pattern, text)
#예상 결과 ['a01020532509@gmail.com', 'i_student@inu.ac.kr', 'kjy@cse.inu.ac.kr']
print(email)

['a01020532509@gmail.com', 'i_student@inu.ac.kr', 'kjy@cse.inu.ac.kr']


## 5주차 수업 내용 리뷰 & 정규 표현식 마무리 -끝-

---

## 6.2 PCA를 이용한 차원 축소



데이터 셋 준비 (우리에게 익숙한 20newsgroups)

In [13]:
from sklearn.datasets import fetch_20newsgroups

#20개의 토픽 중 선택하고자 하는 토픽을 리스트로 생성
categories = ['alt.atheism', 'talk.religion.misc', 'comp.graphics', 'sci.space']

#학습 데이터셋을 가져옴
#메일 내용에서 hint가 되는 부분을 삭제 - 순수하게 내용만으로 분류
newsgroups_train = fetch_20newsgroups(subset='train',
                                      remove=('headers', 'footers', 'quotes'),
                                      categories=categories)
#검증 데이터셋을 가져옴
newsgroups_test = fetch_20newsgroups(subset='test',
                                     remove=('headers', 'footers', 'quotes'),
                                     categories=categories)

X_train = newsgroups_train.data   #학습 데이터셋
y_train = newsgroups_train.target #학습 데이터셋

X_test = newsgroups_test.data     #검증 데이터셋
y_test = newsgroups_test.target   #검증 데이터셋

전처리
* 토큰화, 불용어처리, 스테밍


In [14]:
import nltk

nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [15]:
from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

RegTok = RegexpTokenizer("[\w']{3,}") # 정규포현식으로 토크나이저를 정의
english_stops = set(stopwords.words('english')) #영어 불용어를 가져옴
stemmer = PorterStemmer() # 스테머로 포터스테머 사용

def my_tokenizer(text):
    tokens = RegTok.tokenize(text)
    # stopwords 제외
    words = [word for word in tokens if word not in english_stops]
    # portr stemmer 적용
    features = [stemmer.stem(token) for token in words]
    return features

In [16]:
# 약간의 시간(약 30초)이 소요됩니다.
from sklearn.feature_extraction.text import TfidfVectorizer

#tfidf = TfidfVectorizer(tokenizer=my_tokenizer, max_features=2000, min_df=5, max_df=0.5)
# 토큰화 외에 아무것도 적용하지 않은 채로 확인
tfidf = TfidfVectorizer(tokenizer=my_tokenizer) ### ... 이 부분을 완성하시오
X_train_tfidf = tfidf.fit_transform(X_train) # train set을 변환
X_test_tfidf = tfidf.transform(X_test) # test set을 변환

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [17]:
print('Original tfidf matrix shape:', X_train_tfidf.shape) ### ... 이 부분을 완성하시오

Original tfidf matrix shape: (2034, 20085)


(방법 1) 차원 축소 방법을 적용하기 전의 Logistic Regression을 이용한 분류 성능 확인

In [18]:
from sklearn.linear_model import LogisticRegression

LR_clf = LogisticRegression() #분류기 선언
LR_clf.fit(X_train_tfidf, y_train) # train data를 이용하여 분류기를 학습
print('#Train set score: {:.3f}'.format(LR_clf.score(X_train_tfidf, y_train)))
print('#Test set score: {:.3f}'.format(LR_clf.score(X_test_tfidf, y_test)))

#Train set score: 0.962
#Test set score: 0.761


(방법 2) 차원 축소 방법을 적용한 후의 Logistic Regression을 이용한 분류 성능 확인
* 차원 축소 방법인 PCA를 적용
* 2000차원으로 축소한 뒤 Logistic Regression을 이용한 분류 성능 확인

https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html

explained_variance_ratio_
- Percentage of variance explained by each of the selected components.
- If n_components is not set then all components are stored and the sum of the ratios is equal to 1.0.
- 각각의 새로운 축에 대해, 축소 전 분산에 대한 비율
- 만일 새로운 축들이 원래의 분산을 모두 설명해준다면 explained_variance_ratio_의 합은 1이 될 것임

In [19]:
# 약간의 시간(약 30초)이 소요됩니다.
from sklearn.decomposition import PCA

#  20,085 차원을 2,000 차원으로 축소
pca = PCA(n_components=2000, random_state=7) ### ... 이 부분을 완성하시오
X_train_pca = pca.fit_transform(X_train_tfidf.toarray())
X_test_pca = pca.transform(X_test_tfidf.toarray())

print('Original tfidf matrix shape:', X_train_tfidf.shape)
print('PCA Converted matrix shape:', X_train_pca.shape)
print('Sum of explained variance ratio: {:.3f}'.format(pca.explained_variance_ratio_.sum())) ### ... 이 부분을 완성하시오

Original tfidf matrix shape: (2034, 20085)
PCA Converted matrix shape: (2034, 2000)
Sum of explained variance ratio: 1.000


In [20]:
# PCA로 차원 축소된 X 데이터 이용
LR_clf.fit(X_train_pca, y_train) ### ... 이 부분을 완성하시오
print('#Train set score: {:.3f}'.format(LR_clf.score(X_train_pca, y_train))) ### ... 이 부분을 완성하시오
print('#Test set score: {:.3f}'.format(LR_clf.score(X_test_pca, y_test))) ### ... 이 부분을 완성하시오

#Train set score: 0.962
#Test set score: 0.761


(방법 3) 특성 선택 방식인 라쏘 회귀를 이용
* lasso_clf = LogisticRegression(penalty='l1', solver='liblinear', C=1)
* 차원을 축소

In [21]:
lasso_clf = LogisticRegression(penalty='l1', solver='liblinear', C=1)
lasso_clf.fit(X_train_tfidf, y_train)

print('#Train set score: {:.3f}'.format(lasso_clf.score(X_train_tfidf, y_train)))
print('#Test set score: {:.3f}'.format(lasso_clf.score(X_test_tfidf, y_test)))

import numpy as np
# 계수(coefficient) 중에서 0이 아닌 것들의 개수를 출력
print('#Used features count: {}'.format(np.sum(lasso_clf.coef_ != 0)), 'out of', X_train_tfidf.shape[1])

#Train set score: 0.790
#Test set score: 0.718
#Used features count: 321 out of 20085


(방법 4) 특성 추출 추출 방식인 PCA를 이용
- 차원의 수를 322개로 축소
- pca = PCA(n_components=322, random_state=7)


In [22]:
# 약간의 시간(약 30초)이 소요됩니다.
pca = PCA(n_components=322, random_state=7) ### ... 이 부분을 완성하시오

X_train_pca = pca.fit_transform(X_train_tfidf.toarray()) ### ... 이 부분을 완성하시오
X_test_pca = pca.transform(X_test_tfidf.toarray()) ### ... 이 부분을 완성하시오

# 322차원으로 줄어든 것을 확인
print('PCA Converted X shape:', X_train_pca.shape) ### ... 이 부분을 완성하시오
print('Sum of explained variance ratio: {:.3f}'.format(pca.explained_variance_ratio_.sum()))  ### ... 이 부분을 완성하시오, 분산이 사라짐 절반 가량
# 2,000개에 비해 정보량이 절반 이하로 줄어들었음

LR_clf.fit(X_train_pca, y_train)  # 322차원으로 줄어든 것을 확인
print('#Train set score: {:.3f}'.format(LR_clf.score(X_train_pca, y_train)))  ### ... 이 부분을 완성하시오
print('#Test set score: {:.3f}'.format(LR_clf.score(X_test_pca, y_test))) ### ... 이 부분을 완성하시오
# 라쏘 회귀분석(0.718)을 이용한 특성 선택보다 더 좋은 성능을 보임

PCA Converted X shape: (2034, 322)
Sum of explained variance ratio: 0.438
#Train set score: 0.870
#Test set score: 0.747


(방법 5) 그러면 322개보다 더 줄여서 100개만 이용해보자.
- 여전히 라쏘 회귀분석보다 좋은지 확인


In [23]:
# 약간의 시간(약 30초)이 소요됩니다.
pca = PCA(n_components=100, random_state=7)

X_train_pca = pca.fit_transform(X_train_tfidf.toarray())
X_test_pca = pca.transform(X_test_tfidf.toarray())
print('PCA Converted X shape:', X_train_pca.shape)
print('Sum of explained variance ratio: {:.3f}'.format(pca.explained_variance_ratio_.sum()))

LR_clf.fit(X_train_pca, y_train)
print('#Train set score: {:.3f}'.format(LR_clf.score(X_train_pca, y_train)))
print('#Test set score: {:.3f}'.format(LR_clf.score(X_test_pca, y_test)))
# 라쏘 회귀분석(0.718)을 이용한 특성 선택보다 더 좋은 성능을 보임

PCA Converted X shape: (2034, 100)
Sum of explained variance ratio: 0.211
#Train set score: 0.807
#Test set score: 0.738


# [[실습 6-2-2]] 방법 1 ~ 방법 5의 Test set score를 비교하고, 어떤 방법이 가장 좋은지 결론을 내리시오.
> (가능한 경우에만 작성) 수업 중 설명된 내용을 종합하여 방법 1 ~ 방법 5를 비교 설명해보시오.

- 방법 1의 Test set score: 0.000 (복붙)
- 방법 2의 Test set score: 0.000 (복붙)
- 방법 3의 Test set score: 0.000 (복붙)
- 방법 4의 Test set score: 0.000 (복붙)
- 방법 5의 Test set score: 0.000 (복붙)

비교 결과 작성: ...
> (예시) 0.000 (방법 0) > 0.000 (방법 0) > 0.000 (방법 0) > 0.000 (방법 0) > 0.000 (방법 0). 따라서 방법 0이 가장 좋다.

#방법 1의 Test set score: 0.761

Logistic Regression을 사용한 방법

#방법 2의 Test set score: 0.761

차원 축소 방법을 적용한 후의 Logistic Regression을 이용한 방법, 2000개의 차원을 사용

#방법 3의 Test set score: 0.718

특성 선택 방식인 라쏘 회귀를 이용한 방법

#방법 4의 Test set score: 0.747

특성 추출 추출 방식인 PCA를 이용, 322개의 차원을 사용

#방법 5의 #Test set score: 0.738

위에 방법에서 차원을 322 -> 100개로 줄임

##0.761(방법 1) = 0.761(방법 2) > 0.747(방법 4) > 0.738(방법 5) > 0.718(방법 3)

방법 1과 2는 똑같은 로지스틱 회귀를 이용하였지만 2번의 방법은 PCA를 적용하여 차원을 2000개로 두고 로지스틱 회귀를 이용한 분류를 진행하였다. 하지만 성능 면에서는 동일한 점수를 기록하였으며 라쏘회귀를 이용한 3번의 방법은 가장 낮은 성능 점수를 기록했다. 그리고 PCA를 사용하여 차원을 각각 322개, 100개로 축소한 4,5번의 방법은 라쏘회귀 방식보다는 좋은 성능을 기록하였지만 로지스틱 회귀를 이용하였을 때 보다 낮은 성능을 기록했다.

###점수를 비교해 보았을 때 방법 1과 2가 가장 좋다.

# [파일] -> [다운로드] -> [.ipynb 다운로드]
# 제출 후 **구글 계정 로그아웃** 잘 하시기 바랍니다.